contracts/external/traders/CanonicalOrders.sol

```diff
@@ -0,0 +1,842 @@
+     */
+    function getCurrentPrice(
+        uint256 baseMarket,
+        uint256 quoteMarket
+    )
+        private
+        view
+        returns (uint256)
+    {
+        Monetary.Price memory basePrice = SOLO_MARGIN.getMarketPrice(baseMarket);
+        Monetary.Price memory quotePrice = SOLO_MARGIN.getMarketPrice(quoteMarket);
+        return basePrice.value.mul(PRICE_BASE).div(quotePrice.value);
+    }
+
+    /**
+     * Parses the order, verifies that it is not expired or canceled, and verifies the signature.
+     */
+    function getOrderAndValidateSignature(
+        bytes memory data
+    )
+        private
+        returns (OrderInfo memory)
+    {
+        Require.that(
+            (
+                data.length == NUM_ORDER_BYTES ||
+                data.length == NUM_ORDER_BYTES + NUM_SIGNATURE_BYTES
+            ),
+            FILE,
+            "Cannot parse order from data"
+        );
+
+        // load orderInfo from calldata
+        OrderInfo memory orderInfo;
+        (
+            orderInfo.order,
+            orderInfo.tradeArgs
+        ) = abi.decode(data, (Order, TradeArgs));
+
+        // load tradeArgs from storage if price is zero
+        if (orderInfo.tradeArgs.price == 0) {
+            orderInfo.tradeArgs = g_tradeArgs;
+            g_tradeArgs = TradeArgs({
+                price: 0,
+                fee: 0,
+                isNegativeFee: false
+            });
+        }
+        Require.that(
+            orderInfo.tradeArgs.price != 0,
+            FILE,
+            "TradeArgs loaded price is zero"
+        );
+
+        orderInfo.orderHash = getOrderHash(orderInfo.order);
+        OrderStatus orderStatus = g_status[orderInfo.orderHash];
+
+        // verify valid signature or is pre-approved
+        if (orderStatus == OrderStatus.Null) {
+            bytes memory signature = parseSignature(data);
+            address signer = TypedSignature.recover(orderInfo.orderHash, signature);
+            Require.that(
+                orderInfo.order.makerAccountOwner == signer,
+                FILE,
+                "Order invalid signature",
+                orderInfo.orderHash
+            );
+        } else {
+            Require.that(
+                orderStatus != OrderStatus.Canceled,
+                FILE,
+                "Order canceled",
+                orderInfo.orderHash
+            );
+            assert(orderStatus == OrderStatus.Approved);
+        }
+
+        // verify price
+        TradeArgs memory tradeArgs = orderInfo.tradeArgs;
+        bool validPrice = isBuy(orderInfo.order)
+            ? tradeArgs.price <= orderInfo.order.limitPrice
+            : tradeArgs.price >= orderInfo.order.limitPrice;
+        Require.that(
+            validPrice,
+            FILE,
+            "Fill invalid price"
+        );
+
+        // verify fee
+        bool validFee = isNegativeFee(orderInfo.order)
+            ? (tradeArgs.fee >= orderInfo.order.limitFee) && tradeArgs.isNegativeFee
```

Comment:
Similar to above comment, but we need to pro-rate this accross multiple fills for the proportion of the order being filled. This also requires remembering the fee that has already been paid for an order